ml/src/main/scala/org/apache/spark/ml/h2o/models/H2OMOJOModelBase.scala

```diff
@@ -37,4 +42,30 @@ abstract class H2OMOJOModelBase[T <: SparkModel[T]]
   }
 
   override def write: MLWriter = new H2OMOJOWriter(this, getMojoData)
+
+  protected def applyPredictionUdf(
+      dataset: Dataset[_],
+      udfConstructor: Array[String] => UserDefinedFunction): DataFrame = {
+    val originalDF = dataset.toDF()
+    H2OSchemaUtils.getGetDatasetShape(dataset.schema) match {
+      case DatasetShape.Flat => applyPredictionUdfToFlatDataFrame(originalDF, udfConstructor)
+      case DatasetShape.StructsOnly | DatasetShape.Nested =>
+        val flattenedDF = H2OSchemaUtils.appendFlattenedStructsToDataFrame(originalDF, RowConverter.temporaryColumnPrefix)
+        val flatWithPredictionsDF = applyPredictionUdfToFlatDataFrame(flattenedDF, udfConstructor)
+        flatWithPredictionsDF.schema.foldLeft(flatWithPredictionsDF) { (df, field) =>
+          if (field.name.startsWith(RowConverter.temporaryColumnPrefix)) df.drop(field.name) else df
+        }
+    }
+  }
+
+  private def applyPredictionUdfToFlatDataFrame(
+      flatDataFrame: DataFrame,
+      udfConstructor: Array[String] => UserDefinedFunction): DataFrame = {
+    val relevantColumnNames = flatDataFrame.columns.intersect {
+      getFeaturesCols() ++ getFeaturesCols().map(s => RowConverter.temporaryColumnPrefix + "_" + s)
```

Comment:
Here we are minimizing a set of columns that is passed to udf. This change also tries to include temporary columns. I'm afraid if removed ``` ++ getFeaturesCols().map(s => RowConverter.temporaryColumnPrefix + "_" + s)``` , values from the temporary columns wouldn't get passed into the udf. Am I missing something?

Reply:
What I meant here is that we are not doing any flattening and since we can work directly on the original frame, we don't have to add the prefix to any columns as we don't need temporary columns. I'll have a second look

Reply:
This method is also called from `applyPredictionUdf`. Yeah it's quite confusing. Maybe this method should get features as a parameter.